runtime/parachains/src/hrmp/benchmarking.rs

```diff
@@ -0,0 +1,289 @@
+			parachain: true,
+			genesis_head: vec![1].into(),
+			validation_code: vec![1].into(),
+		},
+	));
+	T::Currency::make_free_balance_be(&id.into_account(), balance);
+}
+
+fn assert_last_event<T: Config>(generic_event: <T as Config>::Event) {
+	let events = frame_system::Pallet::<T>::events();
+	let system_event: <T as frame_system::Config>::Event = generic_event.into();
+	// compare to the last event record
+	let frame_system::EventRecord { event, .. } = &events[events.len() - 1];
+	assert_eq!(event, &system_event);
+}
+
+/// Enumerates the phase in the setup process of a channel between two parachains.
+enum ParachainSetupStep {
+	/// A channel open has been requested
+	Requested,
+	/// A channel open has been requested and accepted.
+	Accepted,
+	/// A channel open has been requested and accepted, and a session has passed and is now open.
+	Established,
+	/// A channel open has been requested and accepted, and a session has passed and is now
+	/// open, and now it has been requested to close down.
+	CloseRequested,
+}
+
+fn establish_para_connection<T: Config>(
+	from: u32,
+	to: u32,
+	until: ParachainSetupStep,
+) -> [(ParaId, crate::Origin); 2]
+where
+	<T as frame_system::Config>::Origin: From<crate::Origin>,
+{
+	let config = Configuration::<T>::config();
+	let deposit: BalanceOf<T> = config.hrmp_sender_deposit.unique_saturated_into();
+	let capacity = config.hrmp_channel_max_capacity;
+	let message_size = config.hrmp_channel_max_message_size;
+
+	let sender: ParaId = from.into();
+	let sender_origin: crate::Origin = from.into();
+
+	let recipient: ParaId = to.into();
+	let recipient_origin: crate::Origin = to.into();
+
+	let output = [(sender, sender_origin.clone()), (recipient, recipient_origin.clone())];
+
+	// Make both a parachain if they are already not.
+	if !Paras::<T>::is_parachain(sender) {
+		register_parachain_with_balance::<T>(sender, deposit);
+	}
+	if !Paras::<T>::is_parachain(recipient) {
+		register_parachain_with_balance::<T>(recipient, deposit);
+	}
+
+	assert_ok!(Hrmp::<T>::hrmp_init_open_channel(
+		sender_origin.clone().into(),
+		recipient,
+		capacity,
+		message_size
+	));
+
+	if matches!(until, ParachainSetupStep::Requested) {
+		return output
+	}
+
+	assert_ok!(Hrmp::<T>::hrmp_accept_open_channel(recipient_origin.into(), sender));
+	if matches!(until, ParachainSetupStep::Accepted) {
+		return output
+	}
+
+	Hrmp::<T>::process_hrmp_open_channel_requests(&Configuration::<T>::config());
+	if matches!(until, ParachainSetupStep::Established) {
+		return output
+	}
+
+	let channel_id = HrmpChannelId { sender, recipient };
+	assert_ok!(Hrmp::<T>::hrmp_close_channel(sender_origin.clone().into(), channel_id));
+	if matches!(until, ParachainSetupStep::CloseRequested) {
+		// NOTE: this is just for expressiveness, otherwise the if-statement is 100% useless.
+		return output
+	}
+
+	output
+}
+
+const PREFIX_0: u32 = 10_000;
+const PREFIX_1: u32 = PREFIX_0 * 2;
+const MAX_UNIQUE_CHANNELS: u32 = 128;
```

Comment:
Where does this number come from?

Reply:
This one does not have a bound anywhere in the codebase, so I left it as a const and u128 should be enough to capture the dependency in the benchmark analysis process. For example, `force_process_hrmp_open ` will iterate over all open channels requests and services them. We know from that

1. it is worse case if all channels have unique sender and receiver.
2. The complexity is indeed `O(2n)` with `n` channels. The main point is to make sure the benchmark system picks up this relation, which it should with any large enough value for `MAX_UNIQUE_CHANNELS`. I reckon if we do 32 or 4096 for it the benchmark should be exactly the same